text/0000-my-feature.md

```diff
@@ -0,0 +1,178 @@
+in the background. This makes sense if what you're doing is truly some kind of
+search query that would need logic to be provided in Ember to also provide
+accurate results but that is not always the case.
+
+# Detailed design
+
+In short, I think it would be useful to allow a `query` object to be provided to
+both `store.findRecord()` and `store.findAll()` inside the already existing
+`options` object. The `query` object would be passed to your adapter and
+serialized into the URL of the request that is made to fetch the resource from a
+server.
+
+## Inclusion of Related Resources
+
+The JSON API spec allows clients to control the [inclusion of related
+resources][fetching-includes] by providing an `include` query parameter when
+requesting a resource.
+
+Including related resources using `store.findRecord()`:
+
+```javascript
+// GET /articles/1?include=comments
+
+const article = this.store.findRecord('article', 1, {
+  query: {
+    include: 'comments'
+  }
+});
+```
+
+Including related resources using `store.findAll()`:
+
+```javascript
+// GET /articles?include=comments
+
+const articles = this.store.findAll('article', {
+  query: {
+    include: 'comments'
+  }
+});
+```
+
+## Sparse Fieldsets
+
+The JSON API spec also allows clients to request that endpoints only return
+specific fields in the response on a per-type basis by including a
+`fields[TYPE]` query parameter.
+
+Here is an example of fetching an article along with its author. In this example
+we're able to specify that we only want the title and body of the article and only
+the name of its author.
+
+```javascript
+// GET /articles/1?include=author&fields[articles]=title,body&fields[users]=name
+
+const article = this.store.findRecord('article', 1, {
+  query: {
+    include: 'author',
+    fields: {
+      articles: 'title,body',
+      users: 'name'
+    }
+  }
+});
+```
+
+Similarly, this can also be done when requesting all articles:
+
+```javascript
+// GET /articles?include=author&fields[articles]=title,body&fields[users]=name
+
+const articles = this.store.findAll('article', {
+  query: {
+    include: 'author',
+    fields: {
+      articles: 'title,body',
+      users: 'name'
+    }
+  }
+});
+```
+
+## This is JSON API Specific
+
+No :simple_smile:. While these examples were for JSON API specific features,
+there is nothing preventing users from passing whatever query parameters they
+might need for features of their own API.
+
+## Implementation
+
+Since `store.findRecord()` and `store.findAll` already accept an `options`
+object, their public APIs will not need to be changed. The private finders that
+they use, however, will need to be modified to pass the `query` (or all of
+`options`) through to their corresponding adapter methods. An example of one of
+the places that would require change is
+[here](https://github.com/emberjs/data/blob/8f83fbf27acdd642cb824e3286832acebdac0da1/packages/ember-data/lib/system/store/finders.js#L19).
```

Comment:
before `adapterOptions` were attached to the `snapshot` we considered passing `adapterOptions` into `findRecord` as an extra argument. In the end we decided to attache it to the snapshot in order to make it work with colaced requests. 

I think we should follow the same pattern here and copy `query` into `adapterOptions.query`. This means 

```js
const articles = this.store.findAll('article', {
  query: {
    include: 'comments'
  }
});
```

would be equivalent to 

```js
const articles = this.store.findAll('article', {
  adapterOptions: {
    query: {
      include: 'comments'
    }
  }
});
```